addons/enemybar/actionTracking.lua

```diff
@@ -0,0 +1,299 @@
+			end
+		end
+	end
+	-- couldn't find the action, let's just give some debug output.
+	if not action_map then
+		action_map = {en='Unknown (id:'..action_id..')'}
+	end 
+
+	tracked_actions[ai.actor_id] = {actor=actor, target=target, ability=action_map, complete=S{4,11,3,5,6}:contains(ai.category), time=os.time()}
+end
+
+function track_debuffs(ai)
+	check_conflicting_debuffs(ai)
+	if S{8,7,6,9}:contains(ai.category) then return end
+
+	for i,t in ipairs(ai.targets) do
+		local target = windower.ffxi.get_mob_by_id(t.id)
+		if target and target.is_npc and not target.charmed then
+
+		    if S{2,252}:contains(t.actions[1].message) then
+		        local spell = ai.param
+		        local effect = res.spells[spell].status
+
+		        if effect then
+		            apply_debuff(target, effect, spell)
+		        end
+		        
+		    -- Non-damaging spells
+		    elseif S{75,236,237,268,270,271}:contains(t.actions[1].message) then
+		        local effect = t.actions[1].param
+		        local spell = ai.param
+
+		        apply_debuff(target, effect, spell)
+		    end
+		end
+	end
+end
+
+function apply_debuff(target, effect, spell)	
+    if not tracked_debuff[target.id] then
+        tracked_debuff[target.id] = {}
+    end
+
+    local overwrites = res.spells[spell].overwrites or {}
+    if not did_overwrite(target, spell, overwrites) then
+        return
+    end
+
+    tracked_debuff[target.id][effect] = {target=target,spell=spell,effect=effect,time=os.time(),duration=res.spells[spell].duration or 0,pos={x=target.x,y=target.y}}
+end
+
+function did_overwrite(target, new, t)
+    if not tracked_debuff[target.id] then
+        return true
+    end
+    
+    for effect, tracked in pairs(tracked_debuff[target.id]) do
+        local old = res.spells[tracked.spell].overwrites or {}
+        
+        -- Check if there isn't a higher priority debuff active
+        if table.length(old) > 0 then
+            for _,v in ipairs(old) do
+                if new == v then
+                    return false
+                end
+            end
+        end
+        
+        -- Check if a lower priority debuff is being overwritten
+        if table.length(t) > 0 then
+            for _,v in ipairs(t) do
+                if tracked.spell == v then
+                    tracked_debuff[target.id][effect] = nil
+                end
+            end
+        end
+    end
+    return true
+end
+
+function check_conflicting_debuffs(ai)
+	local actor = windower.ffxi.get_mob_by_id(ai.actor_id)
+	if not actor then return end
+	if actor.is_npc then
+		-- the actor is an npc, let's check if they're supposed to be asleep/petrified/terror'd
+		local debuffs = tracked_debuff[actor.id]
+		if not debuffs then return end
+
+		for id,debuff in pairs(debuffs) do
+			if S{2,19,7,28}:contains(id) then
+				-- it was inactive, but now it's doing!
+				tracked_debuff[actor.id][id] = nil
+			end
+			if math.abs(debuff.pos.x-actor.x)>0.5 or math.abs(debuff.pos.y-actor.y)>0.5 then
```

Comment:
This and the previous `if` check could be joined with an `or`, but this may actually be better for readability... just pointing it out.